### ISSUE TITLE
Add Valences.{Univalent, Bivalent}.peek() and Traversable.peek()

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#131]
@ggalmazor suggested: "Maybe we could also benefit from something like Try.peek() (as in Stream.peek()) to support side effects on a map/filter pipeline. This would be helpful with objects like the one from your example whose methods don't return themselves (impeding their use on map() phases in the pipeline)."

**Comment by danieldietrich** [comment]
My first impression is that, compared to java.util.stream.Stream, we have (in general) no operation pipeline the elements went through. An instance (e.g. Try, List, Option, ...) of component type `T` would have to take care of an encapsulated optional `Peek` instance, which represents a Consumer of an element. Methods, which perform an operation on one or more elements would have to check, if they have to peek the element before performing the operation (like map, filter, ...).

That seems to be too much overhead and not natural. A possible candidate is the lazy linked list `javaslang.collection.Stream`.

**Comment by ggalmazor** [comment]
I was thinking mainly on legacy apis... 

Let me try with an example (it's not the best... I'll find another). This code is what I'm actually using for hashing strings with SHA256:

```java
String hash = Try.of(() -> MessageDigest.getInstance(DIGEST_ALGORITHM))
    .map(sha256 -> {
      byte[] bytes = Try.of(() -> input.getBytes(ENCODING)).get();
      sha256.update(bytes);
      return sha256;
    })
    .map(MessageDigest::digest)
    .map(digest -> {
      StringBuilder sb = new StringBuilder();
      for (byte aDigest : digest)
        sb.append(String.format("%02x", aDigest));
      return sb.toString();
    })
    .get();
```

With peek(), I could avoid artificially returning the sha256 instance in the first map():

```java
String hash = Try.of(() -> MessageDigest.getInstance(DIGEST_ALGORITHM))
    .peek(sha256 -> {
      byte[] bytes = Try.of(() -> input.getBytes(ENCODING)).get();
      sha256.update(bytes);
    })
    .map(MessageDigest::digest)
    .map(digest -> {
      StringBuilder sb = new StringBuilder();
      for (byte aDigest : digest)
        sb.append(String.format("%02x", aDigest));
      return sb.toString();
    })
    .get();
```

The problem here is that I'm dealing with a legacy, mutable object, and that I'm forced to 'lie' to the Try's api, because I'm not really mapping anything.

I know that I have another better example... but I can't recall :)

Anyway, I recon that it's indeed a very small improvement with an easy workaround...

**Comment by danieldietrich** [comment]
Thank you, great example.

Given an object `o` of type `A<T>` we would define `peek : (T -> void) -> A<T>`, i.e.

```java
interface Peekable<T> {
    Peekable<T> peek(Consumer<T> action);
}
```

Try is a special case. Here we would expect  a `CheckedConsumer` instead.

The semantics of peek depends on the data type. However, in each case it will return the instance it was called on. Additionally the value passed to the consumer is determined as follows:

* Try: the value of Success (or nothing happens)
* Option: the value of Some (or nothing happens)
* Either: the value of either left or right in the context of a left or right projection (or nothing happens in the context of the opposite projection)
* List: the head element or NoSuchElementException on the empty list. Perhaps additionally a `List<T> peekOption(Consumer<Option<T>> action)` which does not throw when called on Nil
* Stream: currently unclear because 1.3.0 may change Stream significantly. Most probably the intermediate operations are deferred in a pipeline of operations. A lazily requested element will walk through the pipeline. Given this, peek will not only called eagerly on the head element but also for all elements of the stream - like in java.util.stream.Stream

Sounds like a plan!

**Comment by ggalmazor** [comment]
yay!

**Comment by danieldietrich** [comment]
:-))

**Comment by danieldietrich** [comment]
Note: `peek` is essentially the same as `forEach`, modulo return type and type of evaluation (lazy vs eager)

**Comment by danieldietrich** [comment]
Instead of creating interfaces `Peekable` and `CheckedPeekable`, `peek()` will walk into `Monad*` and `CheckedMonad*`. This concept of [high cohesion](http://en.wikipedia.org/wiki/Cohesion_(computer_science)) can be found throughout the design Javaslang.

I.e. the checked function `CheckedPredicate`, `CheckedRunnable` and `CheckedSupplier` are part of `Try` (since 1.2.0) because they logically belong to an error handling context - outside of this they make pretty much no sense.

Similar to this and additionally to `flatMap`and `map`, the functions `filter`, `forEach`and `peek` are belonging to a monadic data type. Why? First, peek is very similar to `forEach`, see above. Second, of cause it is a matter of fact that Monads can be seen as [cats in boxes](http://420.thrashbarg.net/3catsinboxes.jpg). We may want to unbox these cats (forEach, peek) or just select only some boxes, depending on the cat inside (filter).

This while idea is a copycat of Scala's monad used in for-comprehensions, which abstracts over cascading for-values (flatMap), which may have if-guards (filter) and finally yields a value (map).

**Comment by danieldietrich** [comment]
Depends on #139

**Comment by danieldietrich** [comment]
Extending the Monad interfaces makes clear, that Tuple does not need to implement Monad.
Furthermore Tuples are the only types which implement Monad2...Monad26.

Therefore Tuple will remove the Monad interface and Monad2-26, Functor2-26 and HigherKinded2-26 will be deleted. (Removing things without touching the business logic is always a good idea)
